### PR TITLE
DMs to show all avatars

### PR DESCRIPTION
**Pull request by WikusKriek**
### Details
Feature request to show up to 8 person Avatars in DMs.

### Fixed Issues

$ https://github.com/Expensify/App/issues/15929#issuecomment-1603765358
PROPOSAL: https://github.com/Expensify/App/issues/15929#issuecomment-1595869495

### Tests

1. Log into the app
2. Create groups of different sizes
3. Verify that avatars display according to [design](https://github.com/Expensify/App/issues/15929#issuecomment-1570177506)
4. Verify that avatars resize Smaller with smaller screen and stack if there are more than 4
5. Verify that avatars resize to Large when less than three person DM on a small screen
6. Create a new workspace and invite 8 or more users, or invite more user to an existing workspace.
7. Click next.
8. Verify that the avatars displayed on the add message screen display according to the [design](https://github.com/Expensify/App/issues/15929#issuecomment-1570177506)
9. Verify that the avatars are of Large size for the add message screen.
- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

Same as Test Steps

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/41372145/9958b0c5-a99a-4a0a-b337-c6d2e5de50de

https://github.com/Expensify/App/assets/41372145/d658e4ab-7f1f-4f7f-95ac-3d19da4dbd40

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/41372145/01a54fb6-7046-402c-a7fa-6b7b29f07070

https://github.com/Expensify/App/assets/41372145/b42e7ddb-5722-4c4a-aee8-b5416ab20783

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/41372145/b93296c7-6acd-4f22-a25d-d5e111ab148c

https://github.com/Expensify/App/assets/41372145/a946ddd8-1cfc-4da9-85cb-251baac82b9b

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/41372145/3a876e9b-b374-4997-994d-7053d39ad21d

https://github.com/Expensify/App/assets/41372145/e9c55fd6-f0d3-48da-b3a0-25bd09a9565c

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/41372145/64b66b62-3719-477f-b90b-f63d0d66dc93

https://github.com/Expensify/App/assets/41372145/3ee49615-507a-400f-9348-7b5eca72f9c6

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/41372145/20f7182c-2f31-4c02-b5b4-1fe0e4c11f06

https://github.com/Expensify/App/assets/41372145/63cbfd9b-830e-478f-9f64-5c7072e9c66b

</details>

### CODE REVIEW

**Comment by shawnborton** [comment]
Nice, this looks good to me. Should we also update the invite to workspace flow where we show avatars as well?

cc @puneetlath for review too

**Comment by puneetlath** [comment]
I only looked at the videos, not the code, but it looks good to me too!

**Comment by puneetlath** [comment]
>Should we also update the invite to workspace flow where we show avatars as well?

On that page it can be a theoretically infinite number of users right? Whereas on the group chat page it's max 8. If so, I think it's fine to leave the invite page as-is. 

Or were you think we'd use the same pattern with up to 8?

**Comment by puneetlath** [comment]
That does look pretty nice.

**Comment by mananjadhav** [comment]
The screencasts look good, I am currently reviewing the code.

**Comment by shawnborton** [comment]
Nice, I like that as well!

**Comment by puneetlath** [comment]
Let's do it then @WikusKriek. And we can double the bounty for the scope increase.

File: src/components/MultipleAvatars.js [hunk 1]
```diff
@@ -49,6 +52,9 @@ const propTypes = {
 
     /** Whether avatars are displayed with the highlighted background color instead of the app background color. This is primarily the case for IOU previews. */
     shouldUseCardBackground: PropTypes.bool,
+
+    /** Prop to limit the amount of avatars displayed horizontally */
+    maxAvatars: PropTypes.number,
```
Comment (mananjadhav) [inline src/components/MultipleAvatars.js:57]:
```suggestion
    maxAvatarsInRow: PropTypes.number,
```

File: src/pages/workspace/WorkspaceInviteMessagePage.js [hunk 2]
```diff
@@ -191,6 +191,7 @@ class WorkspaceInviteMessagePage extends React.Component {
                                 size={CONST.AVATAR_SIZE.LARGE}
                                 icons={OptionsListUtils.getAvatarsForAccountIDs(_.values(this.props.invitedEmailsToAccountIDsDraft), this.props.personalDetails)}
                                 shouldStackHorizontally
+                                shouldStackVertically
```
Comment (mananjadhav) [inline src/pages/workspace/WorkspaceInviteMessagePage.js:194]: Having both `shouldStackHorizontally` and `shouldStackVertically` sound confusing. Can we rename them better based on their use?

File: src/pages/home/report/ReportActionItemCreated.js [hunk 3]
```diff
@@ -76,7 +77,13 @@ function ReportActionItemCreated(props) {
                         accessibilityLabel={props.translate('common.details')}
                         accessibilityRole="button"
                     >
-                        <RoomHeaderAvatars icons={icons} />
+                        <MultipleAvatars
+                            icons={icons}
+                            size={props.isLargeScreenWidth || (icons && icons.length < 3) ? CONST.AVATAR_SIZE.LARGE : CONST.AVATAR_SIZE.MEDIUM}
+                            shouldStackHorizontally
+                            shouldStackVertically={props.isSmallScreenWidth}
+                            maxAvatars={props.isSmallScreenWidth ? 4 : 8}
```
Comment (mananjadhav) [inline src/pages/home/report/ReportActionItemCreated.js:85]: Let's move 4 and 8 to const.

File: src/pages/home/report/ReportActionItemCreated.js [hunk 4]
```diff
@@ -76,7 +77,13 @@ function ReportActionItemCreated(props) {
                         accessibilityLabel={props.translate('common.details')}
                         accessibilityRole="button"
                     >
-                        <RoomHeaderAvatars icons={icons} />
+                        <MultipleAvatars
+                            icons={icons}
+                            size={props.isLargeScreenWidth || (icons && icons.length < 3) ? CONST.AVATAR_SIZE.LARGE : CONST.AVATAR_SIZE.MEDIUM}
```
Comment (mananjadhav) [inline src/pages/home/report/ReportActionItemCreated.js:82]: I am wondering if these should be fixed inside the component? The size? @puneetlath do you think we'll need flexibility of setting this component wise?

File: src/components/MultipleAvatars.js [hunk 5]
```diff
@@ -57,20 +63,34 @@ const defaultProps = {
     secondAvatarStyle: [StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)],
     fallbackIcon: undefined,
     shouldStackHorizontally: false,
+    shouldStackVertically: false,
```
Comment (mananjadhav) [inline src/components/MultipleAvatars.js:66]: Should one of them be true by default?

File: src/components/MultipleAvatars.js [hunk 6]
```diff
@@ -57,20 +63,34 @@ const defaultProps = {
...
 function MultipleAvatars(props) {
+    const [avatarRows, setAvatarRows] = useState([props.icons]);
     let avatarContainerStyles = props.size === CONST.AVATAR_SIZE.SMALL ? [styles.emptyAvatarSmall, styles.emptyAvatarMarginSmall] : [styles.emptyAvatar, styles.emptyAvatarMargin];
     const singleAvatarStyles = props.size === CONST.AVATAR_SIZE.SMALL ? styles.singleAvatarSmall : styles.singleAvatar;
     const secondAvatarStyles = [props.size === CONST.AVATAR_SIZE.SMALL ? styles.secondAvatarSmall : styles.secondAvatar, ...props.secondAvatarStyle];
     const tooltipTexts = props.shouldShowTooltip ? _.pluck(props.icons, 'name') : [''];
 
+    useEffect(() => {
+        if (props.shouldStackVertically && props.icons.length > props.maxAvatars) {
+            const rowSize = Math.min(Math.ceil(props.icons.length / 2), props.maxAvatars);
```
Comment (mananjadhav) [inline src/components/MultipleAvatars.js:85]: Can we move this block to another method? and have some comments there?

File: src/CONST.js [hunk 7]
```diff
@@ -1110,6 +1110,10 @@ const CONST = {
         HEADER: 'header',
         MENTION_ICON: 'mention-icon',
     },
+    AVATAR_ROW_SIZE: {
+        DEFAULT: 4,
+        EIGHT: 8,
```
Comment (mananjadhav) [inline src/CONST.js:1115]:
```suggestion
        LARGE_SCREEN: 8,
```

File: src/components/MultipleAvatars.js [hunk 8]
```diff
@@ -57,20 +63,49 @@ const defaultProps = {
...
+        }
+
+        // Calculate the size of each row
+        const rowSize = Math.min(Math.ceil(props.icons.length / 2), props.maxAvatarsInRow);
+
+        // Slice the icons array into two rows
+        const firstRow = props.icons.slice(rowSize);
+        const secondRow = props.icons.slice(0, rowSize);
+
+        // // Update the state with the two rows as an array
```
Comment (mananjadhav) [inline src/components/MultipleAvatars.js:97]:
```suggestion
        // Update the state with the two rows as an array
```

File: src/pages/home/report/ReportActionItemCreated.js (hunk 4 above)
Comment (puneetlath) [reply src/pages/home/report/ReportActionItemCreated.js:82]: For workspace, it's always shown in the right-hand pane which is always the same-ish size right?

Comment (puneetlath) [reply src/pages/home/report/ReportActionItemCreated.js:82]: That makes sense to me.

**Review by mananjadhav — APPROVED** [review]
Thanks for the changes @WikusKriek. I'll start testing now.

Can also update the test steps to include Workspace invite changes?

**Comment by mananjadhav** [comment]
All on me @WikusKriek. I was away for the weekend, I'll be finishing this one by EOD today.

**Comment by mananjadhav** [comment]
@shawnborton I have been trying to find, if we wanted to change the stacked design even for the large screen? This is how it 
looks web browser with different sizes.

image

Screenshot 2023-07-04 at 12 10 22 AM

Screenshot 2023-07-04 at 12 10 12 AM

**Comment by mananjadhav** [comment]
Thanks for the fix @WikusKriek. Testing again now.

**Comment by mananjadhav** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

web-workspace-invite-avatars

</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/3069065/93e075d0-a19f-458b-8827-4e1a53ea775c

</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>


-------

@WikusKriek this tests well, just one thing is missing, can you upload screencasts for the workspace invite image?

:ribbon: :eyes: :ribbon: C+ Reviewed. @shawnborton @puneetlath all yours.

**Comment by mananjadhav** [comment]
Thanks @WikusKriek. @puneetlath @shawnborton if you folks can take a look.

**Comment by shawnborton** [comment]
I think this feels good to me.

**Review by puneetlath — DISMISSED** [review]

**Comment by puneetlath** [comment]
@WikusKriek FYI you have conflicts. 

@roryabraham do you want to approve as well? You're technically the CME on the issue.

**Review by puneetlath — APPROVED** [review]

**Comment by puneetlath** [comment]
Looks like Rory is OOO, so I'm going to go ahead and merge.

**Pull request merged**